fem/bilininteg_ext.cpp

```diff
@@ -0,0 +1,1581 @@
+// Copyright (c) 2010, Lawrence Livermore National Security, LLC. Produced at
+// the Lawrence Livermore National Laboratory. LLNL-CODE-443211. All Rights
+// reserved. See file COPYRIGHT for details.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability see http://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the GNU Lesser General Public License (as published by the Free
+// Software Foundation) version 2.1 dated February 1999.
+
+#include "../config/config.hpp"
+#include "../general/okina.hpp"
+#include "../linalg/dtensor.hpp"
+
+#include "fem.hpp"
+#include <map>
+#include <cmath>
+#include <algorithm>
+#include "bilininteg.hpp"
+#include "bilininteg_ext.hpp"
+
+using namespace std;
+
+namespace mfem
+{
+
+static const IntegrationRule &DefaultGetRule(const FiniteElement &trial_fe,
+                                             const FiniteElement &test_fe)
+{
+   int order;
+   if (trial_fe.Space() == FunctionSpace::Pk)
+   {
+      order = trial_fe.GetOrder() + test_fe.GetOrder() - 2;
+   }
+   else
+   {
+      // order = 2*el.GetOrder() - 2;  // <-- this seems to work fine too
+      order = trial_fe.GetOrder() + test_fe.GetOrder() + trial_fe.GetDim() - 1;
+   }
+   if (trial_fe.Space() == FunctionSpace::rQk)
+   {
+      return RefinedIntRules.Get(trial_fe.GetGeomType(), order);
+   }
+   return IntRules.Get(trial_fe.GetGeomType(), order);
+}
+
+// PA Diffusion Integrator
+
+// OCCA 2D Assemble kernel
+#ifdef MFEM_USE_OCCA
+static void OccaPADiffusionAssemble2D(const int D1D,
```

Comment:
I may be outvoted on this one, but I think it would be cleaner to move the OCCA integrator methods into their own file.

Reply:
The actual kernels are in `fem/diffusion.okl`